examples/server/server.cpp

```diff
@@ -3073,431 +2890,409 @@ int main(int argc, char **argv)
-                                {"index", i++},
-                                {"object", "embedding"}
-                            };
-                            data.push_back(embedding);
-                        }
-                        json result = format_embeddings_response_oaicompat(body, data);
-                        return res.set_content(result.dump(), "application/json; charset=utf-8");
-                    }
-                }
-                else
-                {
-                    prompt = "";
-                }
+    svr.Post("/tokenize", [&llama](const httplib::Request & req, httplib::Response & res) {
+        res.set_header("Access-Control-Allow-Origin", req.get_header_value("Origin"));
+        const json body = json::parse(req.body);
 
-                // create and queue the task
-                const int task_id = llama.queue_tasks.get_new_id();
-                llama.queue_results.add_waiting_task_id(task_id);
-                llama.request_completion(task_id, { {"prompt", prompt}, { "n_predict", 0}}, false, true, -1);
+        std::vector<llama_token> tokens;
+        if (body.count("content") != 0) {
+            tokens = llama.tokenize(body["content"], false);
+        }
+        const json data = format_tokenizer_response(tokens);
+        return res.set_content(data.dump(), "application/json; charset=utf-8");
+    });
 
-                // get the result
-                task_result result = llama.queue_results.recv(task_id);
-                llama.queue_results.remove_waiting_task_id(task_id);
+    svr.Post("/detokenize", [&llama](const httplib::Request & req, httplib::Response & res) {
+        res.set_header("Access-Control-Allow-Origin", req.get_header_value("Origin"));
+        const json body = json::parse(req.body);
 
-                json data = json::array({json{
-                        {"embedding", json_value(result.result_json, "embedding", json::array())},
-                        {"index", 0},
-                        {"object", "embedding"}
-                    }}
-                );
+        std::string content;
+        if (body.count("tokens") != 0) {
+            const std::vector<llama_token> tokens = body["tokens"];
+            content = tokens_to_str(llama.ctx, tokens.cbegin(), tokens.cend());
+        }
 
-                json root = format_embeddings_response_oaicompat(body, data);
+        const json data = format_detokenized_response(content);
+        return res.set_content(data.dump(), "application/json; charset=utf-8");
+    });
 
-                // send the result
-                return res.set_content(root.dump(), "application/json; charset=utf-8");
-            });
+    svr.Post("/embedding", [&params, &llama](const httplib::Request & req, httplib::Response & res) {
+        res.set_header("Access-Control-Allow-Origin", req.get_header_value("Origin"));
+        if (!params.embedding) {
+            res.status = 400;
```

Comment:
Normally I'd prefer using 403 Forbidden, because it means "the function exists, but it's not accessible". 501 is more like "the function does not exist in the first place".